cmd/kubeadm/app/phases/upgrade/staticpods.go

```diff
@@ -109,13 +115,97 @@ func (spm *KubeStaticPodPathManager) BackupManifestDir() string {
+
+	// Wait for the mirror Pod hash to change; otherwise we'll run into race conditions here when the kubelet hasn't had time to
+	// notice the removal of the Static Pod, leading to a false positive below where we check that the API endpoint is healthy
+	// If we don't do this, there is a case where we remove the Static Pod manifest, kubelet is slow to react, kubeadm checks the
+	// API endpoint below of the OLD Static Pod component and proceeds quickly enough, which might lead to unexpected results.
+	if err := waiter.WaitForStaticPodControlPlaneHashChange(cfg.NodeName, component, beforePodHash); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
 
-	// This string-string map stores the component name and backup filepath (if a rollback is needed).
-	// If a rollback is needed,
+	// Wait for the static pod component to come up and register itself as a mirror pod
+	if err := waiter.WaitForPodsWithLabel("component=" + component); err != nil {
+		return rollbackOldManifests(recoverManifests, err, pathMgr)
+	}
+
+	fmt.Printf("[upgrade/staticpods] Component %q upgraded successfully!\n", component)
+	return nil
+}
+
+// StaticPodControlPlane upgrades a static pod-hosted control plane
+func StaticPodControlPlane(waiter apiclient.Waiter, pathMgr StaticPodPathManager, cfg *kubeadmapi.MasterConfiguration, etcdUpgrade bool) error {
 	recoverManifests := map[string]string{}
 
+	// etcd upgrade is done prior to other control plane components
+	if etcdUpgrade {
+		// Add etcd static pod spec only if external etcd is not configured
+		if len(cfg.Etcd.Endpoints) != 0 {
+			return fmt.Errorf("external etcd cannot be upgraded with kubeadm")
+
+		}
+
+		// Need to check currently used version and version from constants, if differs then uprgade
+		runningEtcdVersion, err := getEtcdRunningVersion()
+		if err != nil {
+			return err
+		}
+		currentEtcdVersion, err := semver.Parse(runningEtcdVersion)
+		if err != nil {
+			return err
+		}
+		upgradeEtcdVersion, err := constants.GetEtcdSupportedVersion(cfg.KubernetesVersion)
+		if err != nil {
+			return err
+		}
+		newEtcdVersion, err := semver.Parse(upgradeEtcdVersion)
+		if err != nil {
+			return err
+		}
+		if currentEtcdVersion.GE(newEtcdVersion) {
+			return fmt.Errorf("etcd upgrade was requested for the same or more recent version")
+		}
+
+		beforeEtcdPodHash, err := waiter.WaitForStaticPodSingleHash(cfg.NodeName, "etcd")
+		if err != nil {
+			return err
+		}
+
+		// Write the updated etcd static Pod manifest into the temporary directory
+		if err := etcdphase.CreateLocalEtcdStaticPodManifestFile(pathMgr.TempManifestDir(), cfg); err != nil {
+			return fmt.Errorf("error creating local etcd static pod manifest file: %v", err)
+		}
+
+		// Perform etcd upgrade using common to all control plane components function
+		if err = performStaticPodUpgrade("etcd", waiter, pathMgr, cfg, beforeEtcdPodHash, recoverManifests); err != nil {
+			return err
+		}
```

Comment:
general comment, I'd really want to perform healthchecks before and after, and rollback if it wasn't upgraded cleanly

Reply:
This PR does not change in any way the existing healthcheck approach.  I can revisit it in the following PR.

Reply:
I can revisit it in the follow up PR